### PR TITLE
bump rust to 1.80.1 / 2024-08-08

### PR DESCRIPTION
**Pull request by yihau**
#### Summary of Changes

- bump rust stable to 1.80.1 from 1.78.0
- bump rust nightly to 2024-08-08 from 2024-05-02
- cherry-pick #2488 
- cherry-pick #2499 + add `no-entrypoint` to the workspace-level lint config.
- run anchor test with their master branch

### CODE REVIEW

**Comment by yihau** [comment]
they released a new patch. will use it

https://blog.rust-lang.org/2024/08/08/Rust-1.80.1.html

**Comment by yihau** [comment]
looking at the latest 3 anchor releases (0.29.0, 0.30.0, 0.30.1), they all depend on time v0.3.29, and Rust 1.80.1 is not happy with this version:

Screenshot 2024-08-14 at 00 04 22

not sure if it's a good idea to just run anchor test with their master branch: https://github.com/anza-xyz/agave/pull/2487/commits/86848e3f40343f3a22115fee9283b47581ed1a36

cc @joncinque

File: .github/workflows/downstream-project-anchor.yml [hunk 1]
```diff
@@ -41,7 +41,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        version: ["v0.29.0", "v0.30.0"]
+        version: ["master"]
```
Comment (brooksprumo) [inline .github/workflows/downstream-project-anchor.yml:44]: I'm not familiar with this part. Is this an intended/needed change?

File: metrics/src/counter.rs [hunk 2]
```diff
@@ -58,7 +58,10 @@ macro_rules! create_counter {
 #[macro_export]
 macro_rules! inc_counter {
     ($name:expr, $level:expr, $count:expr) => {
-        unsafe { $name.inc($level, $count) };
+        #[allow(clippy::macro_metavars_in_unsafe)]
+        unsafe {
+            $name.inc($level, $count)
+        };
```
Comment (brooksprumo) [inline metrics/src/counter.rs:64]: Is the `unsafe` block even needed?

(I looked at this because I wanted to understand the new lint. Here's docs: https://rust-lang.github.io/rust-clippy/master/#/macro_metavars_in_unsafe)

File: .github/workflows/downstream-project-anchor.yml (hunk 1 above)
Comment (yihau) [reply .github/workflows/downstream-project-anchor.yml:44]: yeah, there is some details for it https://github.com/anza-xyz/agave/pull/2487#issuecomment-2286731796

I guess we can either disable it until anchor has a new release or use the current patch.

File: metrics/src/counter.rs (hunk 2 above)
Comment (yihau) [reply metrics/src/counter.rs:64]: yeah, we need the unsafe block. we have some mutable static usage. 🫠  

Screenshot 2024-08-14 at 01 46 14

**Comment by joncinque** [comment]
@yihau I think it's reasonable to use master until a release is tagged with the fix. We might be able to follow along at https://github.com/coral-xyz/anchor/pull/3143

**Comment by yihau** [comment]
thank you for the confirmation @joncinque 🫶  I just submitted an issue as a reminder: #2577

File: metrics/src/counter.rs (hunk 2 above)
Comment (brooksprumo) [reply metrics/src/counter.rs:64]: Ahhh, I see OK.

**Review by brooksprumo — DISMISSED** [review]
👻

**Comment by ryoqun** [comment]
could be done by a follow up pr but seems coverage build is broken...

https://buildkite.com/anza/agave/builds/9332#01914c6e-206a-46bc-bcd5-bd45f56c72e2/6-12191:

```
warning: /tmp/.tmpAk7rfX/profraw/default-82429-4038632013895607840_0_1.profraw: raw profile version mismatch: Profile uses raw profile format version = 10; expected version = 9
PLEASE update this tool to version in the raw profile, or regenerate raw profile with expected version.
warning: /tmp/.tmpAk7rfX/profraw/default-122440-2816328413332674788_0_1.profraw: raw profile version mismatch: Profile uses raw profile format version = 10; expected version = 9
PLEASE update this tool to version in the raw profile, or regenerate raw profile with expected version.
warning: /tmp/.tmpAk7rfX/profraw/default-152543-1154487265959453352_0_1.profraw: raw profile version mismatch: Profile uses raw profile format version = 10; expected version = 9
PLEASE update this tool to version in the raw profile, or regenerate raw profile with expected version.
error: no profile can be merged
17:10:53 [ERROR] Error while executing llvm tools: Failure while running "/usr/local/rustup/toolchains/1.80.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/llvm-profdata" "merge" "-f" "-" "-sparse" "-o" "/tmp/.tmpAk7rfX/0/grcov.profdata"
lcov: /solana/scripts/../target/cov/39418df52/coverage/lcov.info
tar: Removing leading `/solana/ci/../' from member names
artifact: /solana/ci/../target/cov/39418df52/coverage.tar
```

i think we need make sure to use `llvm-profdata` from nightly's `llvm-tools-preview`.

**Comment by yihau** [comment]
let me check the coverage issue 🤔

File: ci/docker/Dockerfile [hunk 3]
```diff
@@ -1,12 +1,13 @@
-FROM ubuntu:20.04
+FROM ubuntu:22.04
```
Comment (brooksprumo) [inline ci/docker/Dockerfile]: This is a pretty big change. I think these should be in their own separate PR.

Comment (yihau) [reply ci/docker/Dockerfile]: oh, actually, it will only affect the tests that use Rust 1.80.1. This means v2.0 and v1.18 won’t be affected by these changes (and old master commits that are using Rust 1.79.0). do you still have concerns if the situation is as I described? to me, it's a really good chance to upgrade those things when we introduce a new rust version haha. just like a new start!

btw, I’m still struggling with the coverage tests 😢 I will keep posting here

Comment (brooksprumo) [reply ci/docker/Dockerfile]: IMO if we can make the ubuntu upgrade *separately* from this PR, we should. Then we keep this PR just about the rust version bump.

Comment (yihau) [reply ci/docker/Dockerfile]: okay! let me try! I think I figured out the coverage issue. need to rebuild the image. will push commits and ping you!

Comment (steviez) [reply ci/docker/Dockerfile]: Good callout Brooks - I'm in favor of bumping the Ubuntu version, but agree that it should be external to this PR

Comment (ryoqun) [reply ci/docker/Dockerfile]: @yihau hmm, maybe i should have been more explicit at my suggestion of this: https://github.com/anza-xyz/agave/pull/2487#issuecomment-2287546356

i think we should use vendorized llvm from the rust-lang, instead of installing the official llvm manually.. this is easier and natually is sync with rust nightly more. in that way, bumping ubuntu (is good thing!) can be separated from this pr.

recent nightly's coverage data is incompatible with llvm 18. it only works now with llvm 19. currently coverage is run under nightly (ie using llvm 19), yet using `llvm-profdata` from _stable_ rustup toolchain (which is still using llvm 18).

```
$ rustup component add --toolchain 1.80.1-x86_64-unknown-linux-gnu llvm-tools-preview
info: downloading component 'llvm-tools'
info: installing component 'llvm-tools'
 31.2 MiB /  31.2 MiB (100 %)  14.8 MiB/s in  2s ETA:  0s
$ rustup component add --toolchain nightly-2024-08-08-x86_64-unknown-linux-gnu llvm-tools-preview
info: downloading component 'llvm-tools'
info: installing component 'llvm-tools'
 32.7 MiB /  32.7 MiB (100 %)  14.7 MiB/s in  2s ETA:  0s

$ /home/ryoqun/.rustup/toolchains/1.80.1-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/llvm-profdata --version
LLVM (http://llvm.org/):
  LLVM version 18.1.7-rust-1.80.1-stable
  Optimized build.
$ /home/ryoqun/.rustup/toolchains/nightly-2024-08-08-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin/llvm-profdata --version
LLVM (http://llvm.org/):
  LLVM version 19.1.0-rust-1.82.0-nightly
  Optimized build.

```

After that, I think we need to just meddle with PATH only running the `grcov`.

Comment (ryoqun) [reply ci/docker/Dockerfile]: (nevertheless, thanks for working on bumping rust. as i like to use latest and greatest rust things (yet not enough time...), i'd like to appreciate your efforts very much.)

**Comment by yihau** [comment]
sorry for the late. I pushed a new commit for fixing coverage test and rebased master for solving conflict. looks all good! it's ready for another review! 🫡

File: define-syscall/Cargo.toml [hunk 4]
```diff
@@ -11,3 +11,8 @@ edition = { workspace = true }
 
 [package.metadata.docs.rs]
 targets = ["x86_64-unknown-linux-gnu"]
+
+[lints.rust]
+unexpected_cfgs = { level = "warn", check-cfg = [
+  'cfg(target_feature, values("static-syscalls"))',
+] }
```
Comment (ryoqun) [inline define-syscall/Cargo.toml:18]: here

File: Cargo.toml [hunk 5]
```diff
@@ -159,6 +159,15 @@ homepage = "https://anza.xyz/"
...
 
+[workspace.lints.rust.unexpected_cfgs]
+level = "warn"
+check-cfg = [
+    'cfg(target_os, values("solana"))',
+    'cfg(feature, values("frozen-abi", "no-entrypoint"))',
+    'cfg(RUSTC_WITH_SPECIALIZATION)',
+    'cfg(RUSTC_WITHOUT_SPECIALIZATION)',
+]
+
```
Comment (ryoqun) [inline Cargo.toml:170]: nit: this can be a follow up pr, but it seems it's prime time for us to finally move lints from `scripts/cargo-clippy-nightly.sh`

for that, i think this formatting is preferred:

```toml
[workspace.lints.rust]
unexpected_cfgs = { level = "warn", check-cfg = [
  ...
] }
```

like this: https://github.com/anza-xyz/agave/pull/2487/files#r1718380391

File: programs/sbf/Cargo.toml [hunk 6]
```diff
@@ -7,6 +7,10 @@ homepage = "https://anza.xyz"
 license = "Apache-2.0"
 edition = "2021"
 
+[workspace.lints.rust.unexpected_cfgs]
```
Comment (ryoqun) [inline programs/sbf/Cargo.toml:10]: nit: same as https://github.com/anza-xyz/agave/pull/2487/files#r1718380621

**Comment by ryoqun** [comment]
seems there's no notable abi-related changes in rust 1.79 and 1.80. that said,

this is a few of my select cool things from the rust change log:

- [Enable frame pointers for the standard library.](https://github.com/rust-lang/rust/pull/122646/)
- [Stabilize associated type bounds (RFC 2289).](https://github.com/rust-lang/rust/pull/122055/)
- [Stabilize inline const {} expressions.](https://github.com/rust-lang/rust/pull/104087/)
- [Stabilize exclusive_range_pattern feature](https://github.com/rust-lang/rust/pull/124459/)
- [LazyCell](https://doc.rust-lang.org/beta/core/cell/struct.LazyCell.html)
- [LazyLock](https://doc.rust-lang.org/beta/std/sync/struct.LazyLock.html)
- [Option::take_if](https://doc.rust-lang.org/beta/std/option/enum.Option.html#method.take_if)

also, this new panic source seems not to be affecting us:
[Panic if PathBuf::set_extension would add a path separator](https://github.com/rust-lang/rust/pull/125070/)

File: scripts/coverage.sh [hunk 7]
```diff
@@ -28,6 +28,14 @@ fi
 # shellcheck source=ci/rust-version.sh
 source "$here/../ci/rust-version.sh" nightly
 
+# Check llvm path
+llvm_profdata="$(find "$(rustc +"$rust_nightly" --print sysroot)" -name llvm-profdata)"
+if [ -z "$llvm_profdata" ]; then
+  echo "Error: couldn't find llvm-profdata. Try installing the llvm-tools component with \`rustup component add llvm-tools-preview --toolchain=$rust_nightly\`"
+  exit 1
+fi
+llvm_path="$(dirname "$llvm_profdata")"
```
Comment (ryoqun) [inline scripts/coverage.sh:37]: :heart:

**Review by ryoqun — APPROVED** [review]
lgtm with nits.

**Comment by yihau** [comment]
🪖
